### PR TITLE
Checkout: Add a partner logo during checkout when partner coupon is present

### PR DESCRIPTION
**Pull request by heydemoura**
#### Changes proposed in this Pull Request

* Add partner logo at the checkout screen of Jetpack products with a partner coupon.
* Context 👉🏻  p1HpG7-dvB-p2#comment-51746
* Summarized objective of this PR 👉🏻  1201265858818217-as-1201439117106183/f

#### Testing instructions

* Accessing [Jetpack Backup Daily checkout](http://calypso.localhost:3000/checkout/jetpack/jetpack_backup_daily?coupon=COUPON_CODE_HERE) should be enough for testing this. Just make sure to have a coupon code to test, you can reach out to the Avalon team to get a testing coupon.
* No need for finishing the order.

Related to #59374

image

### CODE REVIEW

File: client/my-sites/checkout/composite-checkout/components/wp-order-review-line-items.tsx [hunk 1]
```diff
@@ -117,6 +117,7 @@ export function WPOrderReviewLineItems( {
 			{ couponLineItem && (
 				<WPOrderReviewListItem key={ couponLineItem.id }>
 					<NonProductLineItem
+						responseCart={ responseCart }
```
Comment (heydemoura) [inline client/my-sites/checkout/composite-checkout/components/wp-order-review-line-items.tsx]: Passing cart data to _NonProductLineItem_ so that it can tell if there is a partner coupon, since it does not have product information.

File: packages/wpcom-checkout/src/checkout-line-items.tsx [hunk 2]
```diff
@@ -227,6 +239,10 @@ function WPNonProductLineItem( {
 					/>
 				</>
 			) }
+
+			{ responseCart && (
+				<PartnerLogo coupon={ responseCart.coupon } products={ responseCart.products } />
+			) }
```
Comment (heydemoura) [inline packages/wpcom-checkout/src/checkout-line-items.tsx]: Conditionally render _PartnerLogo_ since it depends on cart data.
Passing coupon and products data from cart to the component, so that it renders that proper logo and callout (currently hardcoded to suit IONOS)

File: packages/wpcom-checkout/src/checkout-line-items.tsx [hunk 3]
```diff
@@ -640,6 +656,14 @@ function isCouponApplied( { coupon_savings_integer = 0 }: ResponseCartProduct )
...
 }
 
+function hasPartnerCoupon( coupon: string, products?: ResponseCartProduct[] ): boolean {
+	const productHasSublabel =
+		products && products.find( ( product: ResponseCartProduct ) => getSublabel( product ) );
+	const isPartnerCoupon = coupon.startsWith( 'IONOS_' );
+
+	return ( productHasSublabel && isPartnerCoupon ) || false;
+}
+
```
Comment (heydemoura) [inline packages/wpcom-checkout/src/checkout-line-items.tsx]: Helper that tells whether or not to render the _PartnerLogo_ contents. Currently being used inside the _PartnerLogo_ component, which might seem redundant to have two checks for rendering the same content in different places, the first being at _WPNonProductLineItem_ and the second being inside _PartnerLogo_ component, but in the future this conditionals might scale to support more complex cases.

File: packages/wpcom-checkout/src/checkout-line-items.tsx [hunk 4]
```diff
@@ -707,6 +731,29 @@ function IntroductoryOfferCallout( {
 	return <DiscountCallout>{ text }</DiscountCallout>;
 }
 
+function PartnerLogo( {
```
Comment (heydemoura) [inline packages/wpcom-checkout/src/checkout-line-items.tsx]: _PartnerLogo_ component to render the partner logo and a callout message informing that it is included with the user's hosting plan.

File: packages/wpcom-checkout/src/checkout-line-items.tsx (hunk 3 above)
Comment (heydemoura) [reply packages/wpcom-checkout/src/checkout-line-items.tsx]: The logic for defining _**productHasSubLabel**_ doesn't seem consistent. But according to https://github.com/Automattic/wp-calypso/pull/54811 , Jetpack checkout will only have a single product.

File: packages/wpcom-checkout/src/checkout-line-items.tsx [hunk 5]
```diff
@@ -900,6 +948,13 @@ function WPLineItem( {
 				</LineItemMeta>
 			) }
 
+			{ product && containsPartnerCoupon && (
+				<LineItemMeta>
+					<LineItemSublabelAndPrice product={ product } />
+					<CouponDiscountCallout product={ product } />
+				</LineItemMeta>
+			) }
```
Comment (heydemoura) [inline packages/wpcom-checkout/src/checkout-line-items.tsx:990]: Renders fewer information for when a partner coupon is applied.

File: packages/wpcom-checkout/src/checkout-line-items.tsx (hunk 3 above)
Comment (sirbrillig) [reply packages/wpcom-checkout/src/checkout-line-items.tsx]: Can you explain more about why this logic needs to look at all products in the cart (at least for `PartnerLogo`) for a sublabel and not just the product being rendered?

**Comment by sirbrillig** [comment]
Thank you for leaving such excellent explanations about each of these changes! It does work, but I have a few concerns about the approach.

There's nothing wrong with passing the `ResponseCart` to a line item in principle, but in this case it's an optional property and it's not clear from the API of `NonProductLineItem` when you'd need to pass that argument or not (should we pass it when rendering taxes? probably not). I think the idea is that you should pass it if the item being displayed is a coupon, but `NonProductLineItem` isn't really aware of what type of thing it's rendering, and this change doesn't make it more aware.

In addition, the logic that determines if we should display the logo or not is based on the contents of the cart as a whole, and as such could pretty easily be determined before we even render the coupon line item.

Those things together make me think that a more clear implementation of this would be to decide just before rendering the coupon in `WPOrderReviewLineItems` if we should show the logo. If so, we can render _a different_ component with the logo _instead of_ the coupon line item. Then you don't need nearly as much logic inside the logo line item because it will only be used for this one purpose.

Let me know what you think about that and if it makes sense.

File: packages/wpcom-checkout/src/checkout-line-items.tsx [hunk 6]
```diff
@@ -640,6 +656,14 @@ function isCouponApplied( { coupon_savings_integer = 0 }: ResponseCartProduct )
 	return coupon_savings_integer > 0;
 }
 
+function hasPartnerCoupon( coupon: string, products?: ResponseCartProduct[] ): boolean {
+	const productHasSublabel =
+		products && products.find( ( product: ResponseCartProduct ) => getSublabel( product ) );
```
Comment (sirbrillig) [inline packages/wpcom-checkout/src/checkout-line-items.tsx]: Minor: this could use `some` instead of `find`.

```suggestion
		products && products.some( ( product: ResponseCartProduct ) => !! getSublabel( product ) );
```

File: packages/wpcom-checkout/src/checkout-line-items.tsx [hunk 7]
```diff
@@ -707,6 +731,29 @@ function IntroductoryOfferCallout( {
...
+function PartnerLogo( {
+	coupon,
+	products,
+}: {
+	coupon: string;
+	products?: ResponseCartProduct[];
+} ): JSX.Element | null {
+	const translate = useTranslate();
+
+	if ( hasPartnerCoupon( coupon, products ) ) {
```
Comment (kallehauge) [inline packages/wpcom-checkout/src/checkout-line-items.tsx]: Does `coupon` include partner data as well? A switch condition could potentially prevent future confusion of why "IONOS" is being shown for test coupons or newly onboarded partners :)

File: packages/wpcom-checkout/src/checkout-line-items.tsx [hunk 8]
```diff
@@ -707,6 +731,29 @@ function IntroductoryOfferCallout( {
...
+		return (
+			<LineItemMeta>
+				<div>{ translate( 'Included in your IONOS plan' ) }</div>
+				<div>
+					<img src={ '/calypso/images/jetpack/partners/partner-logo-ionos.png' } alt="IONOS Logo" />
```
Comment (kallehauge) [inline packages/wpcom-checkout/src/checkout-line-items.tsx]: Should we perhaps add max height/width values to the logo to match the suggested design? p1HpG7-dvB-p2#comment-51777

File: packages/wpcom-checkout/src/checkout-line-items.tsx (hunk 7 above)
Comment (heydemoura) [reply packages/wpcom-checkout/src/checkout-line-items.tsx]: It currently does not. In fact it would be a next step to have that data coming from the coupon API somehow (TBD), so that we could potentially serve different partners.

File: packages/wpcom-checkout/src/checkout-line-items.tsx (hunk 8 above)
Comment (heydemoura) [reply packages/wpcom-checkout/src/checkout-line-items.tsx]: I have added max width/height values for the partner logo in this case, but we should reconsider the styling in the future for different partners.

File: packages/wpcom-checkout/src/checkout-line-items.tsx (hunk 7 above)
Comment (kallehauge) [reply packages/wpcom-checkout/src/checkout-line-items.tsx]: We could use the coupon prefix to determine the partner, like we do in `client/jetpack-connect/controller.js::partnerCouponRedirects`. Or even better; refactor the partner detection logic to be a utility function and share it between the two places 😄

File: client/my-sites/checkout/composite-checkout/components/wp-order-review-line-items.tsx (hunk 1 above)
Comment (sirbrillig) [reply client/my-sites/checkout/composite-checkout/components/wp-order-review-line-items.tsx]: As mentioned in my comment above, it seems odd to pass `responseCart` only to _this_ instance of `NonProductLineItem`.

File: packages/wpcom-checkout/src/checkout-line-items.tsx [hunk 9]
```diff
@@ -707,6 +777,21 @@ function IntroductoryOfferCallout( {
...
 
+function PartnerLogo( { className }: { className?: string } ): JSX.Element | null {
+	const translate = useTranslate();
+
+	/* eslint-disable wpcalypso/jsx-classname-namespace */
+	return (
+		<LineItemMeta className={ joinClasses( [ className, 'jetpack-partner-logo' ] ) }>
+			<div>{ translate( 'Included in your IONOS plan' ) }</div>
+			<div className={ 'checkout-line-item__partner-logo-image' }>
+				<img src={ '/calypso/images/jetpack/partners/partner-logo-ionos.png' } alt="IONOS Logo" />
```
Comment (sirbrillig) [inline packages/wpcom-checkout/src/checkout-line-items.tsx]: I know this may still be in-progress, but this URL could potentially be a problem if this component was ever rendered by a package outside of calypso (eg: `mini-cart` when it's [used in wp-admin](https://github.com/Automattic/wp-calypso/pull/52890), even though it currently doesn't import `WPCouponLineItem`).

However, it's not simple to include an image like this inside an npm package, so my suggestions are either:

- Move `PartnerLogo` and `WPCouponLineItem` into calypso (the `client` directory). This is probably the easiest thing unless you think this needs to be used in `mini-cart`.
- Make the image an SVG and move it to the `wpcom-checkout` package, like [some of our payment logos](https://github.com/Automattic/wp-calypso/blob/8b48c93f4f9732481d87e24dd1fd819bc260fd25/packages/wpcom-checkout/src/google-pay-mark.tsx). This is much less efficient but would work.

File: packages/wpcom-checkout/src/checkout-line-items.tsx (hunk 7 above)
Comment (heydemoura) [reply packages/wpcom-checkout/src/checkout-line-items.tsx]: Thanks for the suggestions @kallehauge!
I managed to extract the partner coupon detecting logic to a helper file, exported from the wpcom-checkout package to be used in a higher lever (composite-checkout), now there's much less logic to be handled in `checkout-line-items.tsx`.

File: packages/wpcom-checkout/src/checkout-line-items.tsx (hunk 9 above)
Comment (heydemoura) [reply packages/wpcom-checkout/src/checkout-line-items.tsx]: Hey @sirbrillig , thanks for the suggestions on this. I still don't know much about how and where the `wpcom-checkout` package is being used so I greatly appreciate your effort on explaining your suggestions.

I've converted the IONOS logo image to an SVG and created an SVG component like you suggested. We should serve different partner logos through and API if we decide to replicate this strategy for other partners, if so we will update this for it to work with such API, which will then eliminate the need of storing a partner logo as an SVG on the code base.

**Comment by sirbrillig** [comment]
Looking pretty good to me from a payments perspective! One thing I noticed while testing is that the spacing of the coupon line item is somehow modified by this change so that it no longer is full-width:
coupon-spacing

**Comment by heydemoura** [comment]
> Looking pretty good to me from a payments perspective! One thing I noticed while testing is that the spacing of the coupon line item is somehow modified by this change so that it no longer is full-width: coupon-spacing

I have updated the styles of `CouponLineItem`, making it just a regular block instead of a flex box.

File: packages/wpcom-checkout/src/partner-coupon.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
+import { getSublabel } from './checkout-labels';
+import type { ResponseCartProduct } from '@automattic/shopping-cart';
+
+export function getPartnerCoupon( {
```
Comment (kallehauge) [inline packages/wpcom-checkout/src/partner-coupon.ts:11]: Maybe we should add a comment here that similar functionality exists in `client/jetpack-connect/controller.js` => `partnerCouponRedirects()` (and vice versa) ? 
The `JETPACK_COUPON_PARTNERS` constant it uses could be a singular list containing all supported partners, so we only have to maintain them in one place - it just needs to be an object that's returned instead of a partner string 😄 
But IMO, a comment should be OK for now, so we can get it live 👍

File: packages/wpcom-checkout/src/partner-coupon.ts [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+	coupon,
+	products,
+}: {
+	coupon: string;
+	products?: ResponseCartProduct[];
+} ): boolean {
+	const productHasSublabel =
```
Comment (kallehauge) [inline packages/wpcom-checkout/src/partner-coupon.ts:18]: _Just an observation but IMO it's currently implemented the expected way:_

Technically we have a list of supported partner products in `client/jetpack-connect/constants.js::JETPACK_COUPON_PRESET_MAPPING` - but I don't think it's necessary to check against that list because even if the cart contains an invalid partner product, we still want to show that the coupon belongs to a partner, so they can try and reach out to their support about what's going on.

TL;DR: Looks good 👍

**Review by sirbrillig — COMMENTED** [review]
Thanks for those updates! The code all looks good to me currently, although I haven't tested with a partner coupon to actually see how it looks with the new UI. I'll reach out to the team as suggested to get one, but basically I'd suggest testing three different scenarios to make sure things look good:

1. Checkout at desktop width (as in the screenshots).
2. Checkout at mobile width.
3. A non-checkout page (eg: the plans page) with the mini-cart in the masterbar. (The mini-cart doesn't have to necessarily use this new UI, but we need to make sure that these changes don't adversely affect it.)

File: packages/wpcom-checkout/src/checkout-line-items.tsx [hunk 12]
```diff
@@ -232,6 +258,43 @@ function WPNonProductLineItem( {
...
+	hasDeleteButton?: boolean;
+	removeProductFromCart?: () => void;
+	createUserAndSiteBeforeTransaction?: boolean;
+	isPwpoUser?: boolean;
+	hasPartnerCoupon?: boolean;
+} ): JSX.Element {
+	return (
+		<div
+			className={ joinClasses( [ className, 'coupon-line-item' ] ) }
+			data-partner-coupon={ !! hasPartnerCoupon }
```
Comment (kallehauge) [inline packages/wpcom-checkout/src/checkout-line-items.tsx:283]: _(More of a question than anything - it wouldn't change the final outcome):_ Do we use this solely to render the component differently? If yes, how come we went with an attribute instead of a class here? 😄

**Review by kallehauge — COMMENTED** [review]
I found a few things - but nothing that actually NEEDS to be addressed. I would love the comments to be added, but that's it!

1. A documentation suggestion to avoid build the same code in two places (but so we still can deploy the logic as is) - this could also be documented elsewhere, but I like the idea of it existing in the code so it's guaranteed that engineers sees it 📝 
2. A question about data attribute vs class
3. An observation - but the conclusion is that it works as expected 😄 

Feel free too see this as a technical LGTM 👍 🚢

**Comment by sirbrillig** [comment]
I did actually manage to test with a partner coupon (I just created a fake coupon starting with the `IONOS_` string) and everything looks pretty much as described. It's not quite like the screenshot above, but the variations may be intentional. Here's some screenshots:

Desktop width:
desktop-partner

Mobile width:
mobile-partner

Mini-cart:
mini-partner

**Review by kallehauge — APPROVED** [review]
>It's not quite like the screenshot above, but the variations may be intentional

Thank you for testing the experience as well @sirbrillig (and all your feedback you've given during this PR!) 🏆 

AFAIK the visuals follows the "allowed parameters" we set for the task (aka, the look and feel of the design suggestion exists after some creative freedom to make it fit checkout without compromising the technical components around it) 👍 

---

@heydemoura I've added the comment I asked for in my last review, so you're good to go (ship) when you have time 😄  🚢 

_Both `E2E Tests (desktop) (Web app)` and `E2E Tests (mobile) (Web app)` are failing because of the same unrelated reasons, so feel free to look past these (both fail because of "Start an invalid WordPress" timing out)._

**Review by sirbrillig — APPROVED** [review]

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7136421

Thank you @heydemoura for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7136421).
